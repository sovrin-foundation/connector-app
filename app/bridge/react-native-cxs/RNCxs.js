// @flow
import { NativeModules, Platform } from 'react-native'
import memoize from 'lodash.memoize'
import type {
  Metadata,
  IndyClaimOffer,
  ConnectAgencyResponse,
  RegisterAgencyResponse,
  CreateOneTimeAgentResponse,
  CreatePairwiseAgentResponse,
  AcceptInvitationResponse,
  CxsCredentialOfferResult,
  InitWithGenesisPathConfig,
  VcxProvisionResult,
  VcxProvision,
  CxsInitConfig,
  VcxInitConfig,
  CxsPushTokenConfig,
  VcxCredentialOfferResult,
  VcxCredentialOffer,
  VcxClaimInfo,
  VcxConnectionConnectResult,
  VcxProofRequest,
  WalletTokenInfo,
  PaymentAddress,
} from './type-cxs'
import type { InvitationPayload } from '../../invitation/type-invitation'
import {
  convertAgencyConfigToVcxProvision,
  convertVcxProvisionResultToUserOneTimeInfo,
  convertCxsInitToVcxInit,
  convertCxsPushConfigToVcxPushTokenConfig,
  convertInvitationToVcxConnectionCreate,
  convertVcxConnectionToCxsConnection,
  convertVcxCredentialOfferToCxsClaimOffer,
  paymentHandle,
  getWalletKey,
  convertSovrinTokensToSovrinAtoms,
  convertSovrinAtomsToSovrinTokens,
  convertVcxLedgerFeesToLedgerFees,
} from './vcx-transformers'
import type { UserOneTimeInfo } from '../../store/user/type-user-store'
import type { AgencyPoolConfig } from '../../store/type-config-store'
import type { MyPairwiseInfo } from '../../store/type-connection-store'
import type {
  ClaimOfferPushPayload,
  ClaimPushPayload,
} from '../../push-notification/type-push-notification'
import type {
  WalletHistoryEvent,
  WalletPayload,
} from '../../wallet/type-wallet'
import type { GenericStringObject } from '../../common/type-common'
import type { Passphrase } from '../../backup/type-backup'
import type { GetClaimVcxResult } from '../../claim/type-claim'
import uniqueId from 'react-native-unique-id'
import { smallDeviceMemory } from './type-cxs'
import { secureSet } from '../../services/storage'
import { __uniqueId } from '../../store/type-config-store'
import type { LedgerFeesData } from '../../store/ledger/type-ledger-store'

const { RNIndy } = NativeModules

export async function acceptInvitationVcx(
  connectionHandle: number
): Promise<MyPairwiseInfo> {
  // hard coding connection options to QR type for now, because vcx needs connection options
  // API for vcx assumes that it is running on enterprise side and not from consumer side
  // hence it tries to create connection with connection type.
  // However, our need is not to create a connection but to create a connection instance
  // with existing invitation. So, for now for any invitation type QR or SMS
  // we are hard coding connection option to QR
  const connectionOptions = { connection_type: 'QR', phone: '' }
  const result: string = await RNIndy.vcxAcceptInvitation(
    connectionHandle,
    JSON.stringify(connectionOptions)
  )
  // TODO:KS Remove below API call once sdk team returns pairwise info in above api
  // above call does not return pairwise did information, but we need pairwise info
  // to store that information and have those details available while making a connection
  // we have to make an extra call to get pairwise info
  const serializedConnection: string = await serializeConnection(
    connectionHandle
  )

  const vcxConnection: { data: VcxConnectionConnectResult } = JSON.parse(
    serializedConnection
  )

  return convertVcxConnectionToCxsConnection(vcxConnection.data)
}

export async function updatePushTokenVcx(pushTokenConfig: CxsPushTokenConfig) {
  return await RNIndy.vcxUpdatePushToken(
    JSON.stringify(convertCxsPushConfigToVcxPushTokenConfig(pushTokenConfig))
  )
}

export async function deleteConnection(connectionHandle: number) {
  return await RNIndy.deleteConnection(connectionHandle)
}

export async function resetVcx(removeData: boolean): Promise<boolean> {
  // we would remove above reset method and rename this method to reset
  // once we have integration available with vcx
  const result: boolean = await RNIndy.reset(true)

  return result
}

export async function vcxShutdown(deletePool: boolean): Promise<boolean> {
  const result: Number = await RNIndy.shutdownVcx(deletePool)

  return true
}

export async function getColor(imagePath: string): Promise<Array<string>> {
  return RNIndy.getColor(imagePath)
}

export async function sendTokenAmount(
  tokenAmount: string,
  recipientWalletAddress: string
): Promise<boolean> {
  const sovrinAtoms = convertSovrinTokensToSovrinAtoms(tokenAmount)
  return RNIndy.sendTokens(paymentHandle, sovrinAtoms, recipientWalletAddress)
}

export async function createOneTimeInfo(
  agencyConfig: AgencyPoolConfig
): Promise<UserOneTimeInfo> {
  const walletPoolName = await getWalletPoolName()
  const vcxProvisionConfig: VcxProvision = await convertAgencyConfigToVcxProvision(
    agencyConfig,
    walletPoolName
  )
  const provisionVcxResult: string = await RNIndy.createOneTimeInfo(
    JSON.stringify(vcxProvisionConfig)
  )
  const provisionResult: VcxProvisionResult = JSON.parse(provisionVcxResult)
  return convertVcxProvisionResultToUserOneTimeInfo(provisionResult)
}

export async function init(
  config: CxsInitConfig,
  fileName: string
): Promise<boolean> {
  const genesis_path: string = await RNIndy.getGenesisPathWithConfig(
    config.poolConfig,
    fileName
  )

  const initConfig = {
    ...config,
    genesis_path,
  }
  const walletPoolName = await getWalletPoolName()
  const vcxInitConfig: VcxInitConfig = await convertCxsInitToVcxInit(
    initConfig,
    walletPoolName
  )
  const initResult: boolean = await RNIndy.init(JSON.stringify(vcxInitConfig))

  return initResult
}

// TODO:KS Need to rename this to something like walletInit
export async function simpleInit(): Promise<boolean> {
  const walletPoolName = await getWalletPoolName()
  const wallet_key = await getWalletKey()
  const initConfig = {
    wallet_name: walletPoolName.walletName,
    wallet_key,
  }
  const initResult: boolean = await RNIndy.init(JSON.stringify(initConfig))
  return initResult
}

export const getWalletPoolName = memoize(async function() {
  const appUniqueId = await uniqueId()
  await secureSet(__uniqueId, appUniqueId)
  const walletName = `${appUniqueId}-cm-wallet`
  // Not sure why, but VCX is applying validation check on pool name
  // they don't like alphanumeric or _, so we have to remove "-"
  // from our guid that we generated
  const strippedAppUniqueId = appUniqueId.replace(/\-/g, '')
  const poolName = `${strippedAppUniqueId}cmpool`
  return {
    walletName,
    poolName,
  }
})

export async function createConnectionWithInvite(
  invitation: InvitationPayload
): Promise<number> {
  const { invite_details } = convertInvitationToVcxConnectionCreate(invitation)
  const connectionHandle: number = await RNIndy.createConnectionWithInvite(
    invitation.requestId,
    JSON.stringify(invite_details)
  )

  return connectionHandle
}

export async function serializeConnection(
  connectionHandle: number
): Promise<string> {
  const serializedConnection: string = await RNIndy.getSerializedConnection(
    connectionHandle
  )

  return serializedConnection
}

// cache Promise of serializedString so that we don't call Bridge method again
export const getHandleBySerializedConnection = memoize(async function(
  serializedConnection: string
): Promise<number> {
  const connectionHandle: number = await RNIndy.deserializeConnection(
    serializedConnection
  )

  return connectionHandle
})

export async function downloadClaimOffer(
  connectionHandle: number,
  sourceId: string,
  messageId: string
): Promise<CxsCredentialOfferResult> {
  const {
    credential_offer,
    credential_handle,
  }: VcxCredentialOfferResult = await RNIndy.credentialCreateWithMsgId(
    sourceId,
    connectionHandle,
    messageId
  )

  const {
    credential_offer: vcxCredentialOffer,
    price,
  }: { credential_offer: VcxCredentialOffer, price: number } = JSON.parse(
    credential_offer
  )

  vcxCredentialOffer.price = price
    ? convertSovrinAtomsToSovrinTokens(price)
    : null

  return {
    claimHandle: credential_handle,
    claimOffer: convertVcxCredentialOfferToCxsClaimOffer(vcxCredentialOffer),
  }
}

export async function serializeClaimOffer(
  claimHandle: number
): Promise<string> {
  const serializedClaimOffer: string = await RNIndy.serializeClaimOffer(
    claimHandle
  )

  return serializedClaimOffer
}

export const getClaimHandleBySerializedClaimOffer = memoize(async function(
  serializedClaimOffer: string
): Promise<number> {
  const claimHandle: number = await RNIndy.deserializeClaimOffer(
    serializedClaimOffer
  )

  return claimHandle
})

export async function sendClaimRequest(
  claimHandle: number,
  connectionHandle: number,
  paymentHandle: number
): Promise<void> {
  return await RNIndy.sendClaimRequest(
    claimHandle,
    connectionHandle,
    paymentHandle
  )
}

export async function downloadProofRequest(
  sourceId: string,
  connectionHandle: number,
  messageId: string
): Promise<VcxProofRequest> {
  const {
    proofHandle,
    proofRequest,
  }: {
    proofHandle: number,
    proofRequest: string,
  } = await RNIndy.proofCreateWithMsgId(sourceId, connectionHandle, messageId)

  const vcxProofRequest: VcxProofRequest = JSON.parse(proofRequest)

  return {
    ...vcxProofRequest,
    proofHandle,
  }
}

export async function getWalletBalance(): Promise<string> {
  const { balance_str: balance }: WalletTokenInfo = await getWalletTokenInfo()

  return convertSovrinAtomsToSovrinTokens(balance)
}

export async function getWalletAddresses(): Promise<string[]> {
  const { addresses }: WalletTokenInfo = await getWalletTokenInfo()

  // TODO:KS For now we don't need to store any other data on our react-native
  // when we need to store all data, then we would return only addresses
  // for now we are just returning addresses and ignoring anything else
  return addresses.map((address: PaymentAddress) => address.address)
}

export async function getWalletHistory(): Promise<WalletHistoryEvent[]> {
  const walletHistoryData = [
    {
      id: 'asd',
      senderAddress:
        'sov:ksudgyi8f98gsih7655hgifuyg79s89s98ydf98fg7gks8fjhkss8f030',
      action: 'Withdraw',
      tokenAmount: '5656',
      timeStamp: 'Tue, 04 Aug 2015 12:38:41 GMT',
    },
    {
      id: 'kld',
      senderName: 'Sovrin Foundation',
      senderAddress:
        'sov:ksudgyi8f98gsih7655hgifuyg79s89s98ydf98fg7gks8fjhkss8f030',
      action: 'Purchase',
      tokenAmount: '10000',
      timeStamp: 'Tue, 04 Aug 2015 14:38:41 GMT',
    },
  ]

  return new Promise.resolve(walletHistoryData)
}

export async function encryptWallet({
  encryptedFileLocation,
  recoveryPassphrase,
}: {
  encryptedFileLocation: string,
  recoveryPassphrase: Passphrase,
}): Promise<number> {
  const exportHandle: number = await RNIndy.exportWallet(
    encryptedFileLocation,
    recoveryPassphrase.hash
  )

  return exportHandle
}

export async function decryptWalletFile(
  walletPath: string,
  decryptionKey: string
): Promise<number> {
  const { walletName } = await getWalletPoolName()
  const wallet_key = await getWalletKey()

  const config = JSON.stringify({
    wallet_name: walletName,
    wallet_key,
    exported_wallet_path: walletPath,
    backup_key: decryptionKey,
  })
  const importHandle: number = await RNIndy.decryptWalletFile(config)

  return importHandle
}

export async function copyToPath(
  uri: string,
  destPath: string
): Promise<number> {
  return await RNIndy.copyToPath(uri, destPath)
}

export async function updateClaimOfferState(claimHandle: number) {
  const updatedState: number = await RNIndy.updateClaimOfferState(claimHandle)

  return updatedState
}

export async function getClaimOfferState(claimHandle: number): Promise<number> {
  const state: number = await RNIndy.getClaimOfferState(claimHandle)

  return state
}

export async function getClaimVcx(
  claimHandle: number
): Promise<GetClaimVcxResult> {
  const vcxClaimResult: string = await RNIndy.getClaimVcx(claimHandle)
  const vcxClaim: VcxClaimInfo = JSON.parse(vcxClaimResult)
  const { credential, credential_id } = vcxClaim

  if (!credential || !credential_id) {
    throw new Error('credential not found in vcx')
  }

  return {
    claimUuid: credential_id,
    claim: credential,
  }
}

export async function exportWallet(wallet: WalletPayload): Promise<number> {
  const exportHandle: number = await RNIndy.exportWallet(
    wallet.walletPath,
    wallet.encryptionKey
  )

  return exportHandle
}

export async function setWalletItem(
  key: string,
  value: string
): Promise<number> {
  return await RNIndy.setWalletItem(key, value)
}

export async function getWalletItem(key: string): Promise<string> {
  const response: string = await RNIndy.getWalletItem(key)
  if (response) {
    const itemValue = JSON.parse(response)
    const { value } = itemValue

    if (!value) {
      throw new Error('cannot get value')
    }

    return value
  } else {
    return response
  }
}

export async function deleteWalletItem(key: string): Promise<number> {
  return await RNIndy.deleteWalletItem(key)
}

export async function updateWalletItem(
  key: string,
  data: string
): Promise<number> {
  return await RNIndy.updateWalletItem(key, data)
}

export function exitAppAndroid() {
  if (Platform.OS === 'android') {
    RNIndy.exitAppAndroid()
  }
}

export async function getMatchingCredentials(
  proofHandle: number
): Promise<string> {
  // TODO:KS Add a transformer here, for now we are just passing through
  return await RNIndy.proofRetrieveCredentials(proofHandle)
}

export async function generateProof(
  proofHandle: number,
  selectedCredentials: string,
  selfAttestedAttributes: string
): Promise<void> {
  return await RNIndy.proofGenerate(
    proofHandle,
    selectedCredentials,
    selfAttestedAttributes
  )
}

export async function sendProof(
  proofHandle: number,
  connectionHandle: number
): Promise<void> {
  return await RNIndy.proofSend(proofHandle, connectionHandle)
}

export async function proofCreateWithRequest(
  sourceId: string,
  proofRequest: string
): Promise<number> {
  return await RNIndy.proofCreateWithRequest(sourceId, proofRequest)
}

export async function proofSerialize(proofHandle: number): Promise<string> {
  return await RNIndy.proofSerialize(proofHandle)
}

export async function proofDeserialize(
  serializedProofRequest: string
): Promise<number> {
  return await RNIndy.proofDeserialize(serializedProofRequest)
}

export async function downloadMessages(
  messageStatus: string,
  uid_s: ?string,
  pwdids: string
): Promise<string> {
  return await RNIndy.downloadMessages(messageStatus, uid_s, pwdids)
}

export async function updateMessages(
  messageStatus: string,
  pwdidsJson: string
): Promise<number> {
  return await RNIndy.updateMessages(messageStatus, pwdidsJson)
}

export async function getWalletTokenInfo(): Promise<WalletTokenInfo> {
  const paymentHandle = 0
  const tokenInfo: string = await RNIndy.getTokenInfo(paymentHandle)
  return JSON.parse(tokenInfo)
}

export async function createPaymentAddress(seed: ?string) {
  return RNIndy.createPaymentAddress(seed)
}

export async function getLedgerFees(): Promise<LedgerFeesData> {
  const fees: string = await RNIndy.getLedgerFees()

  return convertVcxLedgerFeesToLedgerFees(fees)
}

export const checkIfAnimationToUse = memoize(function() {
  return Platform.OS === 'android'
    ? RNIndy.totalMemory / smallDeviceMemory < 1 ? true : false
    : false
})

export async function getBiometricError(): Promise<string> {
  if (Platform.OS === 'ios') {
    return await RNIndy.getBiometricError()
  }
  return ''
}
