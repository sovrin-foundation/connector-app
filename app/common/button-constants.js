// @flow
export const DENY = 'Deny'
export const CONNECT = 'Connect'
