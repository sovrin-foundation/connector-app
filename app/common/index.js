// @flow
export * from './button-constants'
export * from './message-constants'
export * from './route-constants'
export * from './secure-storage-constants'
export function noop() {}
