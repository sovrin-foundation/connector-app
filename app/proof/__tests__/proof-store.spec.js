// @flow
import renderer from 'react-test-renderer'
import { put, call, select, take } from 'redux-saga/effects'
import { initialTestAction } from '../../common/type-common'
import proofReducer, {
  proofSuccess,
  generateProofSaga,
  convertPreparedProofToRequestedAttributes,
  convertSelfAttestedToIndySelfAttested,
  proofFail,
  convertIndyPreparedProofToAttributes,
  getProof,
} from '../proof-store'
import {
  acceptProofRequest,
  missingAttributesFound,
  proofRequestAutoFill,
  sendProof,
  proofRequestShowStart,
} from '../../proof-request/proof-request-store'
import { generateProof } from '../../bridge/react-native-cxs/RNCxs'
import {
  proofRequest,
  proofRequest10Attributes,
  proof,
  proofWithMissingAttributes,
  proofWith10Attributes,
  originalProofRequestData,
  originalProofRequestData10Attributes,
  originalProofRequestDataMissingAttribute,
  preparedProof,
  homeAddressPreparedProof,
  preparedProofWithMissingAttribute,
  homeAddressPreparedProofWithMissingAttribute,
  selfAttestedAttributes,
  selfAttestedAttributes1,
  poolConfig,
  originalProofRequestDataWithSpaces,
} from '../../../__mocks__/static-data'
import {
  getOriginalProofRequestData,
  getProofRequestPairwiseDid,
  getPoolConfig,
} from '../../store/store-selector'
import {
  USER_SELF_ATTESTED_ATTRIBUTES,
  UPDATE_ATTRIBUTE_CLAIM,
} from '../type-proof'
import { PROOF_REQUEST_SHOW_START } from '../../proof-request/type-proof-request'

describe('Proof Store', () => {
  const remoteDid = proofRequest.payloadInfo.remotePairwiseDID
  const uid = proofRequest.payloadInfo.uid

  let initialState
  let afterProofSuccess
  beforeEach(() => {
    initialState = proofReducer(undefined, initialTestAction())
    afterProofSuccess = proofReducer(initialState, proofSuccess(proof, uid))
  })

  it('should match snapshot for proof received action', () => {
    expect(afterProofSuccess).toMatchSnapshot()
  })

  it('should convert prepared proof to indy proof format', () => {
    expect(
      convertPreparedProofToRequestedAttributes(
        preparedProof,
        originalProofRequestData
      )
    ).toMatchSnapshot()
  })

  it('should find missing attributes if any missing', () => {
    expect(
      convertPreparedProofToRequestedAttributes(
        preparedProofWithMissingAttribute,
        originalProofRequestDataMissingAttribute
      )
    ).toMatchSnapshot()
  })

  it('match snapshot while converting SelfAttestedAttributes to Indy format', () => {
    expect(
      convertSelfAttestedToIndySelfAttested(selfAttestedAttributes)
    ).toMatchSnapshot()
    expect(
      convertSelfAttestedToIndySelfAttested(selfAttestedAttributes1)
    ).toMatchSnapshot()
  })

  it('fn:convertPreparedProofToRequestedAttributes', () => {
    expect(
      convertIndyPreparedProofToAttributes(
        homeAddressPreparedProof,
        originalProofRequestData.requested_attributes
      )
    ).toMatchSnapshot()
    expect(
      convertIndyPreparedProofToAttributes(
        homeAddressPreparedProof,
        originalProofRequestData10Attributes.requested_attributes
      )
    ).toMatchSnapshot()

    expect(
      convertIndyPreparedProofToAttributes(
        homeAddressPreparedProof,
        originalProofRequestDataWithSpaces.requested_attributes
      )
    ).toMatchSnapshot()
  })

  it('store update correctly if proof fails', () => {
    expect(
      proofReducer(
        initialState,
        proofFail(uid, {
          code: 'TEST-CODE',
          message: 'test error message',
        })
      )
    ).toMatchSnapshot()
  })

  // TODO:KS Fix these tests before July 25
  xit('generate proof saga should generate proof', () => {
    const gen = generateProofSaga(getProof(uid))
    expect(gen.next().value).toEqual(select(getOriginalProofRequestData, uid))

    expect(gen.next(originalProofRequestData).value).toEqual(
      select(getPoolConfig)
    )

    const preparedProofJson = JSON.stringify(homeAddressPreparedProof)
    const requestedClaimsJson = {
      self_attested_attributes: {},
      requested_attrs: convertPreparedProofToRequestedAttributes(
        homeAddressPreparedProof,
        originalProofRequestData
      )[0],
      requested_predicates: {},
    }

    const requestedAttributes = convertIndyPreparedProofToAttributes(
      {
        ...homeAddressPreparedProof,
        self_attested_attrs: {},
      },
      originalProofRequestData.requested_attributes
    )

    expect(gen.next(preparedProofJson).value).toEqual(
      put(proofRequestAutoFill(uid, requestedAttributes))
    )

    expect(gen.next().value).toEqual(select(getProofRequestPairwiseDid, uid))

    expect(gen.next(remoteDid).value).toEqual(take(UPDATE_ATTRIBUTE_CLAIM))

    expect(
      gen.next({
        type: UPDATE_ATTRIBUTE_CLAIM,
        requestedAttrsJson: requestedClaimsJson.requested_attrs,
      }).value
    ).toEqual(put(sendProof(uid)))

    expect(gen.next(JSON.stringify(proof)).value).toEqual(
      put(proofSuccess(proof, uid))
    )

    expect(gen.next().value).toEqual(put(acceptProofRequest(uid)))

    expect(gen.next().done).toBe(true)
  })

  // TODO:KS Fix this tests before July 25
  xit('generate proof saga should work fine with missing attributes', () => {
    const gen = generateProofSaga(getProof(uid))
    expect(gen.next().value).toEqual(select(getOriginalProofRequestData, uid))

    expect(gen.next(originalProofRequestDataMissingAttribute).value).toEqual(
      select(getPoolConfig)
    )

    const preparedProofJson = JSON.stringify(
      homeAddressPreparedProofWithMissingAttribute
    )
    const [
      requestedAttributesJson,
      missingAttributes,
    ] = convertPreparedProofToRequestedAttributes(
      homeAddressPreparedProofWithMissingAttribute,
      originalProofRequestDataMissingAttribute
    )

    expect(gen.next(preparedProofJson).value).toEqual(
      put(missingAttributesFound(missingAttributes, uid))
    )

    expect(gen.next(preparedProofJson).value).toEqual(
      take(USER_SELF_ATTESTED_ATTRIBUTES)
    )

    const requestedClaimsJson = {
      self_attested_attributes: convertSelfAttestedToIndySelfAttested(
        selfAttestedAttributes
      ),
      requested_attrs: requestedAttributesJson,
      requested_predicates: {},
    }

    const requestedAttributes = convertIndyPreparedProofToAttributes(
      {
        ...homeAddressPreparedProofWithMissingAttribute,
        self_attested_attrs: { ...selfAttestedAttributes },
      },
      originalProofRequestDataMissingAttribute.requested_attributes
    )

    expect(
      gen.next({
        type: USER_SELF_ATTESTED_ATTRIBUTES,
        selfAttestedAttributes,
        uid,
      }).value
    ).toEqual(put(proofRequestAutoFill(uid, requestedAttributes)))

    expect(gen.next().value).toEqual(select(getProofRequestPairwiseDid, uid))

    expect(gen.next(remoteDid).value).toEqual(take(UPDATE_ATTRIBUTE_CLAIM))

    expect(
      gen.next({
        type: UPDATE_ATTRIBUTE_CLAIM,
        requestedAttrsJson: requestedClaimsJson.requested_attrs,
      }).value
    ).toEqual(put(sendProof(uid)))

    expect(gen.next(JSON.stringify(proof)).value).toEqual(
      put(proofSuccess(proof, uid))
    )

    expect(gen.next().value).toEqual(put(acceptProofRequest(uid)))

    expect(gen.next().done).toBe(true)
  })

  it('should reset proof store, if RESET action is raised', () => {
    expect(proofReducer(afterProofSuccess, { type: 'RESET' })).toMatchSnapshot()
  })

  it('proof store should reset for a given proof request if PROOF_REQUEST_SHOW_START is raised', () => {
    const proofStateAfterFail = proofReducer(
      initialState,
      proofFail(uid, {
        code: 'TEST-CODE',
        message: 'test error message',
      })
    )
    expect(
      proofReducer(proofStateAfterFail, proofRequestShowStart(uid))
    ).toMatchSnapshot()
  })
})
