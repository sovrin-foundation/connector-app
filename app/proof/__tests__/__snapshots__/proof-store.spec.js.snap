// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Proof Store fn:convertPreparedProofToRequestedAttributes 1`] = `
Array [
  Array [
    Object {
      "claimUuid": "claim::ea03d8ca-eeb4-4944-b7d6-5abcf4503d73",
      "cred_info": Object {
        "cred_info": Object {
          "attrs": Object {
            "Address 1": "Address 1",
          },
          "cred_def_id": "V4SGRU86Z58d6TV7PBUe6f:3:CL:24:tag1",
          "referent": "claim::ea03d8ca-eeb4-4944-b7d6-5abcf4503d73",
          "schema_id": "V4SGRU86Z58d6TV7Pf:2:slKljrSQ80tCQ40F:33089",
        },
      },
      "data": "Address 1",
      "key": "attr1_uuid",
      "label": "Address 1",
    },
  ],
  Array [
    Object {
      "claimUuid": "claim::6a0f42b4-1210-4bdb-ad53-10ed765276b5",
      "cred_info": Object {
        "cred_info": Object {
          "attrs": Object {
            "Address 2": "Address 2",
          },
          "cred_def_id": "V4SGRU86Z58d6TV7PBUe6f:3:CL:24:tag1",
          "referent": "claim::6a0f42b4-1210-4bdb-ad53-10ed765276b5",
          "schema_id": "V4SGRU86Z58d6TV7Pf:2:slKljrSQ80tCQ40F:33089",
        },
      },
      "data": "Address 2",
      "key": "attr2_uuid",
      "label": "Address 2",
    },
  ],
]
`;

exports[`Proof Store fn:convertPreparedProofToRequestedAttributes 2`] = `
Array [
  Array [
    Object {
      "claimUuid": "claim::ea03d8ca-eeb4-4944-b7d6-5abcf4503d73",
      "cred_info": Object {
        "cred_info": Object {
          "attrs": Object {
            "Address 1": "Address 1",
          },
          "cred_def_id": "V4SGRU86Z58d6TV7PBUe6f:3:CL:24:tag1",
          "referent": "claim::ea03d8ca-eeb4-4944-b7d6-5abcf4503d73",
          "schema_id": "V4SGRU86Z58d6TV7Pf:2:slKljrSQ80tCQ40F:33089",
        },
      },
      "data": "Address 1",
      "key": "attr1_uuid",
      "label": "Address 1",
    },
  ],
  Array [
    Object {
      "claimUuid": "claim::6a0f42b4-1210-4bdb-ad53-10ed765276b5",
      "cred_info": Object {
        "cred_info": Object {
          "attrs": Object {
            "Address 2": "Address 2",
          },
          "cred_def_id": "V4SGRU86Z58d6TV7PBUe6f:3:CL:24:tag1",
          "referent": "claim::6a0f42b4-1210-4bdb-ad53-10ed765276b5",
          "schema_id": "V4SGRU86Z58d6TV7Pf:2:slKljrSQ80tCQ40F:33089",
        },
      },
      "data": "Address 2",
      "key": "attr2_uuid",
      "label": "Address 2",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "label": "Address 3",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "label": "Address 4",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "label": "Address 5",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "label": "Address 6",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "label": "Address 7",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "label": "Address 8",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "label": "Address 9",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "label": "Address 10",
    },
  ],
]
`;

exports[`Proof Store fn:convertPreparedProofToRequestedAttributes 3`] = `
Array [
  Array [
    Object {
      "claimUuid": "claim::ea03d8ca-eeb4-4944-b7d6-5abcf4503d73",
      "cred_info": Object {
        "cred_info": Object {
          "attrs": Object {
            "Address 1": "Address 1",
          },
          "cred_def_id": "V4SGRU86Z58d6TV7PBUe6f:3:CL:24:tag1",
          "referent": "claim::ea03d8ca-eeb4-4944-b7d6-5abcf4503d73",
          "schema_id": "V4SGRU86Z58d6TV7Pf:2:slKljrSQ80tCQ40F:33089",
        },
      },
      "data": "Address 1",
      "key": "attr1_uuid",
      "label": "A d d r e s s    1",
    },
  ],
  Array [
    Object {
      "claimUuid": "claim::6a0f42b4-1210-4bdb-ad53-10ed765276b5",
      "cred_info": Object {
        "cred_info": Object {
          "attrs": Object {
            "Address 2": "Address 2",
          },
          "cred_def_id": "V4SGRU86Z58d6TV7PBUe6f:3:CL:24:tag1",
          "referent": "claim::6a0f42b4-1210-4bdb-ad53-10ed765276b5",
          "schema_id": "V4SGRU86Z58d6TV7Pf:2:slKljrSQ80tCQ40F:33089",
        },
      },
      "data": "Address 2",
      "key": "attr2_uuid",
      "label": "  Ad  dress 2 ",
    },
  ],
]
`;

exports[`Proof Store match snapshot while converting SelfAttestedAttributes to Indy format 1`] = `
Object {
  "attr2_uuid": "male",
}
`;

exports[`Proof Store match snapshot while converting SelfAttestedAttributes to Indy format 2`] = `
Object {
  "attr2_uuid": "male",
  "attr3_uuid": "190",
}
`;

exports[`Proof Store proof store should reset for a given proof request if PROOF_REQUEST_SHOW_START is raised 1`] = `
Object {
  "pid123": undefined,
}
`;

exports[`Proof Store should convert prepared proof to indy proof format 1`] = `
Array [
  Object {
    "attr1_uuid": Array [
      "claim::ea03d8ca-eeb4-4944-b7d6-5abcf4503d73",
      true,
    ],
    "attr2_uuid": Array [
      "claim::6a0f42b4-1210-4bdb-ad53-10ed765276b5",
      true,
    ],
  },
  Array [],
]
`;

exports[`Proof Store should find missing attributes if any missing 1`] = `
Array [
  Object {
    "attr1_uuid": Array [
      "claim::ea03d8ca-eeb4-4944-b7d6-5abcf4503d73",
      true,
    ],
    "attr3_uuid": Array [
      "claim::6a0f42b4-1210-4bdb-ad53-10ed765276b5",
      true,
    ],
  },
  Array [
    Object {
      "key": "attr2_uuid",
      "name": "address2",
    },
  ],
]
`;

exports[`Proof Store should match snapshot for proof received action 1`] = `
Object {
  "pid123": Object {
    "aggregated_proof": Object {
      "c_hash": "31",
      "c_list": Array [
        Array [
          36,
        ],
      ],
    },
    "proofs": Object {
      "claim::f760213b-e719-47fb-9669-b72c5c410e8c": Object {
        "issuer_did": "issuerDid",
        "proof": Object {
          "non_revoc_proof": null,
          "primary_proof": Object {
            "eq_proof": Object {
              "a_prime": "13",
              "e": "51",
              "m": Object {},
              "m1": "67",
              "m2": "60",
              "revealed_attrs": Object {
                "name": "11",
                "sex": "59",
              },
              "v": "13",
            },
            "ge_proofs": Array [],
          },
        },
        "schema_seq_no": 36,
      },
    },
    "requested_proof": Object {
      "predicates": Object {},
      "revealed_attrs": Object {
        "attr1_uuid": Array [
          "claim::f760213b-e719-47fb-9669-b72c5c410e8c",
          "Alex",
          "11",
        ],
        "attr2_uuid": Array [
          "claim::f760213b-e719-47fb-9669-b72c5c410e8c",
          "male",
          "59",
        ],
      },
      "self_attested_attrs": Object {},
      "unrevealed_attrs": Object {},
    },
  },
}
`;

exports[`Proof Store should reset proof store, if RESET action is raised 1`] = `Object {}`;

exports[`Proof Store store update correctly if proof fails 1`] = `
Object {
  "pid123": Object {
    "error": Object {
      "code": "TEST-CODE",
      "message": "test error message",
    },
  },
}
`;
